### ISSUE TITLE
Trigger IX-F import when network sets allow_ixp_update to "yes"

### ISSUE THREAD

**Issue by arnoldnipper** [peeringdb/peeringdb#499]
Currently, the IX-F JSON importer runs daily at midnight UTC. Even when increasing the frequency the expectation might be that when setting `allow_ixp_update` to "yes", changes are incorporated instantly as it would be when configured manually.

**Comment by ccaputo** [comment]
I'd imagine the update that happens when a network enables allow_ixp_update is restricted to just that network, rather than a full run of the importer.  Further, I could see this using cached JSON data rather than re-downloading the data based on this user change.  Ie., limit JSON downloads to whatever the IX-F import cron frequency is.

**Comment by arnoldnipper** [comment]
@ccaputo to be more precise, it would be limited to rerun the importer for those IXes where the network is connected. However downloading/checking for fresh data is imho a MUST.

**Comment by ccaputo** [comment]
@arnoldnipper, I disagree since the list of IXPs a network is connected to may not be known until a full run is completed.  Some may discovered as part of the run.

**Comment by arnoldnipper** [comment]
Bummer, @ccaputo  :) ... however, I can think of the use case that network is newly connected to IX and wants the data to be visible instantly. However, cached data of IX will not have data for network yet.

**Comment by ccaputo** [comment]
@arnoldnipper I hear you.  A challenge is make sure there is no DoS aspect of a ton of allow_ixp_update enablings all at once resulting in the same tasks on the PeeringDB server.  There may need to be a hybrid solution of having known IXes for a network be updated based on cached JSON data while new IXes for a network get discovered as part of the next periodic cron run.  The timeout for cached JSON data could be as low as, say, a minute.

**Comment by ccaputo** [comment]
Posted the above before finishing: Another thought is that there become parallel tables for the IX-F data vs. user-entered data and the allow_ixp_update switch just toggles which table is used.  That would enable the data to become instantly available, as long as it was known at the last importer run.

**Comment by arnoldnipper** [comment]
Or triggers are queued to avoid dossing. 2 or up to 5 minutes still may mean instantly. Then the next regular download is again done after $interval time.

**Comment by arnoldnipper** [comment]
@vegu did we take this issue into account when designing the new process?

**Comment by arnoldnipper** [comment]
@ccaputo and @vegu, shouldn't this be much easier now with the new importer? And @pc, please make up your mind.

**Comment by ccaputo** [comment]
I don't know about easier, but I imagine this could be queued like the /cp/ method of queuing imports.  Needs @peeringdb/pc decision, I think.

**Comment by arnoldnipper** [comment]
@ccaputo, with the new importer you already know about all potential IX-F JSON induced changes. Hence, when a network sets `allow_ixp_update` to "yes" you would only have to rescan this list and perform the actions needed. And no need to queue anything.

**Comment by ccaputo** [comment]
> @ccaputo, with the new importer you already know about all potential IX-F JSON induced changes. Hence, when a network sets `allow_ixp_update` to "yes" you would only have to rescan this list and perform the actions needed. And no need to queue anything.
> 
> But anyway, @peeringdb/pc, please comment

Sorry - I was confused about network-based "Allow IXP Update" versus IX-based "Enable IX-F Import".  That said, I think we'd need to hear from @vegu about what is involved with making this work, with respect to data previously loaded from an IX-F Import.

**Comment by arnoldnipper** [comment]
@peeringdb/pc, while with the old importer there was no knowing what was imported, the new importer stores this information in a separate table. Hence, it would now be relatively easy to go through this table and act on the suggestions when a network sets `allow_ixp_update` to `true`. That way the network immediately sees the changes and would not have to wait for the next import run. Result: increased user experience.

What do you think? Please +1 :)

**Comment by arnoldnipper** [comment]
@peeringdb/pc please comment. IMO users massively would benefit from this feature

**Comment by arnoldnipper** [comment]
This issue would need just another +1, @peeringdb/pc. Low hanging fruit as well

**Comment by mcmanuss8** [comment]
+1

**Comment by grizz** [comment]
I'm a fan of this idea, but we don't currently have a job queue at all, so we would have to change the IX-F importer to poll for changes periodically. Would fit in nicely with #962

**Comment by arnoldnipper** [comment]
> ... but we don't currently have a job queue at all, so we would have to change the IX-F importer to poll for changes periodically

Not sure why you need a job queue. Wouldn't you only have to go through the IX-F imported `netixlan` objects with net_id the same as the network which made the change of `allow_ixp_update` to yes?

Anyway, nice to see that you like the idea. Implementation left to you :)

**Comment by arnoldnipper** [comment]
@vegu, what's left to do?

**Comment by vegu** [comment]
@arnoldnipper 

I think we have all the tools to make this work now.  

> Not sure why you need a job queue. Wouldn't you only have to go through the IX-F imported netixlan objects with net_id the same as the network which made the change of allow_ixp_update to yes

This is straight forward way to do this - go through all the network's suggestions and apply them when the network changes its `allow_ixp_update` to `yes`. However it comes with the caveat that this would not be functional for new networks until the ix-f importer has ran again, since no suggestions are created for networks that do not exist in PeeringDB at the time of import.

I think we can just run the importer using cached json against a single asn - we already do this when generating the network preview. 

Need to double check performance difference between previewing changes and actually committing the import, but if performance is an issue we now also have a queue to use.
